### PR TITLE
Set cookies and crumb

### PR DESCRIPTION
**Pull request by dpguthrie**
- Initialize session with required cookies
- Create larger list of rotating user-agents with related headers
- Added webdriver-manager for selenium functionality
- Changed from flit to poetry
- Added additional package files that will help with testing and development
- Added function in base client class to retrieve a crumb and then add as a query parameter

### CODE REVIEW

**Comment by ValueRaider** [comment]
I quickly tried this but get exception:
`yq.Ticker('AMZN', username='redacted', password='redacted')`
>   File ".../yahooquery/yahooquery/login.py", line 27, in __init__
>     self.driver = webdriver.Chrome(
>                   ^^^^^^^^^^^^^^^^^
> TypeError: WebDriver.__init__() got an unexpected keyword argument 'service'

`selenium=3.141.0  webdriver_manager=3.8.6`

**Comment by dpguthrie** [comment]
> selenium=3.141.0 webdriver_manager=3.8.6

Can you try upgrading selenium?

**Comment by ValueRaider** [comment]
Selenium 4 just generates a different exception:
> requests.exceptions.RetryError: HTTPSConnectionPool(host='query2.finance.yahoo.com', port=443): Max retries exceeded with url: /v1/test/getcrumb (Caused by ResponseError('too many 500 error responses'))

I prefer v3 because calls with v4 lag for 2 minutes, I've no idea why

**Comment by dpguthrie** [comment]
Do you get the same error when not passing in a username and password?

**Comment by ValueRaider** [comment]
Without username & password, same "max retries" error and now with either version of Selenium. Note: does not happen with yahooquery 2.3.2 (latest official)

File: yahooquery/base.py [hunk 1]
```diff
@@ -947,24 +948,13 @@ def symbols(self):
     def symbols(self, symbols):
         self._symbols = _convert_to_list(symbols)
         
-    def get_crumb(self):
+    def set_crumb(self):
         response = self.session.get('https://query2.finance.yahoo.com/v1/test/getcrumb')
```
Comment (ericwbzhang) [inline yahooquery/base.py:952]: Hi Shall we enable proxy here as well?

Comment (dpguthrie) [reply yahooquery/base.py:952]: The proxies are already set on the session when it's being [initialized](https://github.com/dpguthrie/yahooquery/blob/cb5b2bd5a7e351ba29414aca3259815e8b4ead6c/yahooquery/utils/__init__.py#L57).

**Comment by ValueRaider** [comment]
Same behaviour, "Max retries" errors. What are other people experiencing?

**Comment by dpguthrie** [comment]
> Same behaviour, "Max retries" errors. What are other people experiencing?

That’s weird. I have it working locally and I have this code running in the streamlit app as well - https://yahooquery.streamlit.app/

**Comment by dpguthrie** [comment]
@ValueRaider I’m wondering if it has something to do with where you’re based. I’m naively navigating to finance.yahoo.com to retrieve cookies - but I believe you’re based in the UK and would actually navigate to uk.finance.yahoo.com. Curious if you pulled this code down and just changed this if it would start working for you.

**Comment by ValueRaider** [comment]
I don't quite understand where I would make the change. But to check if worth investigating, I logged in with Firefox Debugger active and don't see UK-specific URLs, but the same as you e.g. `https://query2.finance.yahoo.com/v1/test/getcrumb`

File: yahooquery/utils/__init__.py [hunk 2]
```diff
@@ -69,31 +161,29 @@ def _init_session(session=None, **kwargs):
...
-#     driver.quit()
-#     return cookies
+        session, crumb = setup_session_with_cookies_and_crumb(session)
+    return session, crumb
 
 
+def setup_session_with_cookies_and_crumb(session: Session):
+    headers = {**random.choice(HEADERS), **addl_headers}
+    session.headers = headers
+    response = session.get('https://finance.yahoo.com', hooks={'response': get_crumb})
```
Comment (dpguthrie) [inline yahooquery/utils/__init__.py:171]: @ValueRaider This is the place where you would modify to (I'm guessing) `https://uk.finance.yahoo.com`

Comment (dpguthrie) [reply yahooquery/utils/__init__.py:171]: I wonder if it has something to do with accepting their cookies. I tried navigating to the uk site and got their consent screen.

image

I’m guessing without accepting that then YF is unable to actually set the appropriate cookies.

Comment (dpguthrie) [reply yahooquery/utils/__init__.py:171]: Also, this code doesn’t use selenium. It’s only used to login to YF (and not even sure that’s working right now as it looks like they’ve added a captcha)

Comment (ValueRaider) [reply yahooquery/utils/__init__.py:171]: The getcrumb `response.text` contains string "HTTP Status 403 - Forbidden". Is that typical of Yahoo expecting a cookie?

Comment (ValueRaider) [reply yahooquery/utils/__init__.py:171]: Sorry wrong part of code, that 403 happens in `setup_session_with_cookies_and_crumb()` on `https://finance.yahoo.com`

Comment (dpguthrie) [reply yahooquery/utils/__init__.py:171]: That's really surprising!  I'm not sure what would make that a forbidden request just to the YF home page.

**Comment by cmjordan42** [comment]
Hey y'all, interested in this one as I currently use `all_modules` in an occasional batch run that computes a ton of single security analytics. Sounds like we're still pretty far from having a solution we like here, and that maybe Yahoo is deliberately discouraging it. Given that, I'm probably gonna switch over to a bunch of individual targeted calls to get what's in `all_modules` on Sunday.

**Comment by dpguthrie** [comment]
> Hey y'all, interested in this one as I currently use `all_modules` in an occasional batch run that computes a ton of single security analytics. Sounds like we're still pretty far from having a solution we like here, and that maybe Yahoo is deliberately discouraging it. Given that, I'm probably gonna switch over to a bunch of individual targeted calls to get what's in `all_modules` on Sunday.

@cmjordan42 Where are you based?  My somewhat uneducated theory right now is that this would work for anyone based in the U.S, which is why it works for me and it works inside of the streamlit app.  If you are U.S. based, you could take this branch for a spin and see if you're able to access the `all_modules` property in that way.

Another somewhat uneducated guess - I think the solution to this is to almost have country/region specific headers that map to what they would be if that person navigated to the YF home page on a browser.  I think what's happening right now is instead of going directly to the home page, it first navigates to `consent.yahoo.com` and has you accept cookies, which you can't do using simply requests.

**Comment by cmjordan42** [comment]
I'm in US - EST. I'd be happy to test drive if it would be helpful.

**Comment by cmjordan42** [comment]
Just tested with this branch, it appears to work fine. Definitely an improvement even if some regions have more complicated issues. May as well publish it to get `all_modules` working for some people.

However, it occasionally fails with something a response along the lines of:

"For input string: \"-91000.0000000002\""

**Comment by cmjordan42** [comment]
And that appears to be transient; if you request the same security again, it may work. Not sure if that transience is on the `yahooquery` side or the Yahoo side.

**Comment by dpguthrie** [comment]
@ValueRaider, @cmjordan42 Would you mind giving this another go?  Made some changes last night that hopefully both handles errors but also provides a fallback option with selenium (if it's installed) to retrieve cookies/crumb.

**Comment by ValueRaider** [comment]
No change, and I tried changing URL and header parameters. These are my browser headers btw:
``` JSON
{
    "Host": "query1.finance.yahoo.com",
    "User-Agent": "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0",
    "Accept": "*/*",
    "Accept-Language": "en-GB,en;q=0.5",
    "Accept-Encoding": "gzip, deflate, br",
    "Content-Type": "text/plain",
    "Origin": "https://finance.yahoo.com",
    "Referer": "https://finance.yahoo.com/"
}
```

**Comment by dpguthrie** [comment]
@ValueRaider Is [this](https://github.com/dpguthrie/yahooquery/pull/205/files#diff-e23edea32223392398eaca1c039d0d57bafea93ad7236b4b3cf1cedf35e70285R200) where the error is happening (the explicit request to retrieve the crumb)?

**Comment by ValueRaider** [comment]
Does this work for you in USA? Works for me in Europe. https://stackoverflow.com/questions/76065035/yahoo-finance-v7-api-now-requiring-cookies-python

**Pull request merged**